### PR TITLE
Replace Adapter with Converter.

### PR DESCRIPTION
**Pull request by NightlyNexus**
Force consumers to serialize to Strings and remove the hook into the
SharedPreferences.

This is totally unsolicited nonsense from me.
I figured I'd just propose it because I already had the code.
Definitely close if it's not to your preference.

### CODE REVIEW

**Comment by JakeWharton** [comment]
While I generally like this approach, one disadvantage here is that you're forced to use String. I can't, for example, convert an `Instant` to a `long` and vise versa. That may be uncommon enough to simply live with though.

File: rx-preferences/src/main/java/com/f2prateek/rx/preferences2/CustomTypeStringAdapter.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package com.f2prateek.rx.preferences2;
+
+import android.content.SharedPreferences;
+import android.support.annotation.NonNull;
+
+import static com.f2prateek.rx.preferences2.Preconditions.checkNotNull;
+
+final class CustomTypeStringAdapter<T> implements RealPreference.Adapter<T> {
```
Comment (f2prateek) [inline rx-preferences/src/main/java/com/f2prateek/rx/preferences2/CustomTypeStringAdapter.java]: I think the name `CustomTypeStringAdapter` could be improved. Possibly something involving `Object` and/or `Converter` — e.g. `ObjectAdapter`, `ConverterAdapter`, `ObjectConverterAdapter`?

File: rx-preferences/src/main/java/com/f2prateek/rx/preferences2/Preference.java [hunk 2]
```diff
@@ -9,18 +8,10 @@
...
-
-    /**
-     * Store non-null {@code value} for {@code key} in {@code editor}.
-     * <p>
-     * Note: Implementations <b>must not</b> call {@code commit()} or {@code apply()} on
-     * {@code editor}.
-     */
-    void set(@NonNull String key, @NonNull T value, @NonNull SharedPreferences.Editor editor);
+  interface Converter<T> {
+    @NonNull T deserialize(@NonNull String serialized);
```
Comment (f2prateek) [inline rx-preferences/src/main/java/com/f2prateek/rx/preferences2/Preference.java]: Let's add some javadocs here.

**Comment by f2prateek** [comment]
Agreed - `Instant` to `long` is a much better example (I was thinking `List<String>` to `Set<String>`).

But I do think it make client code easier to write (no chance of accidentally calling `commit`, `apply`) so I think the tradeoff is worth it.

File: rx-preferences/src/main/java/com/f2prateek/rx/preferences2/CustomTypeStringAdapter.java (hunk 1 above)
Comment (NightlyNexus) [inline rx-preferences/src/main/java/com/f2prateek/rx/preferences2/CustomTypeStringAdapter.java]: Yep! I'll try `ConverterAdapter`.

File: rx-preferences/src/main/java/com/f2prateek/rx/preferences2/Preference.java (hunk 2 above)
Comment (NightlyNexus) [inline rx-preferences/src/main/java/com/f2prateek/rx/preferences2/Preference.java]: Yep, I'm not tied to the Converter method names, either.

**Comment by NightlyNexus** [comment]
Sorry for the delay (now I'm on vacation).
I definitely understand the hesitation; I just haven't had a use case outside of strings, so I'm also good continuing my own fork.
I'll follow up tonight with changes. Thanks!

**Comment by onomated** [comment]
@f2prateek @NightlyNexus  This new interface breaks some functionality for one of my use cases.  Particularly with the fact that I no longer have access to the preference key.
I have an EncryptedPreferenceAdapter that generally worked as follows with the previous Adapter interface:
```java
    @Override
    public Optional<T> get(@NonNull String key, @NonNull SharedPreferences preferences) {
            final String serialized = preferences.getString(key, null);
            if (serialized == null)
                return Optional.empty();
            final String decrypted = Obfuscator.decryptFromStorage(key, serialized);
            return Optional.ofNullable(decrypted);
    }

    @Override
    public void set(@NonNull String key, @NonNull Optional<T> optionalValue, @NonNull SharedPreferences.Editor editor) {
            if (optionalValue.isPresent()) {
                editor.putString(key, Obfuscator.encryptForStorage(key, optionalValue.get()));
            }
            else {
                editor.putString(key, null);
            }
    }
```
Can Converters be updated to provide the preference key?  Or can adapters be resurrected since there may be apps out in the wild that have adapters serializing to non-string types as @JakeWharton pointed out?

**Pull request merged**